File: pkgs/backend/src/groupsum_catalog_api/schemas/catalog.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class ApiModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class CatalogOverview(ApiModel):
    kind: Literal["catalog_overview"]
    generated_at: datetime | None = None
    counts: dict[str, int]


class MetricPoint(ApiModel):
    observed_at: datetime | None = None
    value: float


class CommitPoint(ApiModel):
    date: str
    count: int


class RepositorySummary(ApiModel):
    id: str
    owner: str
    name: str
    url: str
    description: str | None = None
    route: str
    package_count: int = 0
    resource_count: int = 0
    release_count: int = 0
    metrics: dict[str, float] = Field(default_factory=dict)
    history: dict[str, list[MetricPoint]] = Field(default_factory=dict)
    commit_activity: list[CommitPoint] = Field(default_factory=list)
    observed_at: datetime | None = None


class PackageSummary(ApiModel):
    id: str
    ecosystem: str
    name: str
    registry_url: str
    route_key: str | None = None
    route: str
    package_kind: str
    publication_status: str | None = None
    latest_version: str | None = None
    release_count: int = 0
    dependency_count: int = 0
    dependent_count: int = 0
    observed_at: datetime | None = None


class TypedResourceSummary(ApiModel):
    id: str
    resource_type: str
    route_key: str | None = None
    title: str
    url: str
    summary: str | None = None
    route: str
    repository_owner: str | None = None
    repository_name: str | None = None
    observed_at: datetime | None = None
    type_label: str | None = None
    resource_family: str | None = None
    icon_key: str | None = None


class TechnologySummary(ApiModel):
    id: str
    slug: str
    name: str
    category: str | None = None
    description: str | None = None
    record_count: int = 0
    route: str


class CatalogCollection(ApiModel):
    kind: str
    resource_kind: Literal["repository", "package", "resource", "technology"]
    count: int
    page: int = 1
    page_size: int = 50
    page_count: int = 1
    facets: dict[str, dict[str, int]] = Field(default_factory=dict)
    generated_at: datetime | None = None
    records: list[RepositorySummary | PackageSummary | TypedResourceSummary | TechnologySummary]


class CatalogMember(ApiModel):
    kind: str
    resource_type: str | None = None
    item: dict[str, Any]
    graph: dict[str, Any] | None = None
    linked_sections: list[dict[str, Any]] = Field(default_factory=list)
    parent: dict[str, Any] | None = None
    implementation: dict[str, Any] = Field(default_factory=dict)
    governance: dict[str, Any] = Field(default_factory=dict)
    legal: dict[str, Any] = Field(default_factory=dict)
    related_records: list[dict[str, Any]] = Field(default_factory=list)

File: pkgs/frontend/src/api/catalog.ts
import { catalogFetch } from "./client";
import type { RecordCollectionPageModel } from "./catalog.generated";

export * from "./catalog.generated";

export type CatalogRecord = Record<string, unknown> & { id: string };
export type RepositorySummary = CatalogRecord;
export type PackageSummary = CatalogRecord;
export type CatalogResourceSummary = CatalogRecord;
export type TechnologySummary = CatalogRecord;
export type CatalogCollection = {
  generated_at?: string | null;
  count: number;
  page: number;
  page_count: number;
  page_size: number;
  records: CatalogRecord[];
  facets: Record<string, Record<string, number>>;
  aggregates: Record<string, number>;
};
export type CatalogMember = {
  item: CatalogRecord;
  resource_type?: string;
  parent?: CatalogRecord | null;
  graph?: unknown;
  linked_sections?: unknown[];
  related_records?: CatalogRecord[];
  governance?: Record<string, unknown>;
  legal?: Record<string, unknown>;
  implementation?: Record<string, unknown>;
};

export type CatalogDataset = "repositories" | "packages" | "resources" | "technologies";
export type CatalogQuery = {
  page: number;
  page_size: number;
  q?: string;
  owner?: string;
  ecosystem?: string;
  publication_status?: string;
  resource_type?: string;
  repository_owner?: string;
  sort?: string;
};

export async function getRecordCollectionPageModel(
  collection: "products" | "portfolio",
  signal?: AbortSignal,
): Promise<RecordCollectionPageModel> {
  const response = await catalogFetch(`/api/v1/${collection}`, {
    signal,
    headers: { Accept: "application/json" },
    cache: "default",
  });
  if (!response.ok) throw new Error(`${collection} collection response ${response.status}`);
  return response.json() as Promise<RecordCollectionPageModel>;
}

function routeRecord(dataset: CatalogDataset, record: CatalogRecord): CatalogRecord {
  if (dataset === "repositories") return { ...record, kind: "repository", route: String(record.route || `/catalog/repositories/${encodeURIComponent(String(record.owner || ""))}/${encodeURIComponent(String(record.name || record.id))}`) };
  if (dataset === "packages") return { ...record, kind: "package", route: String(record.route || `/catalog/packages/${encodeURIComponent(String(record.ecosystem || "package"))}/${encodeURIComponent(String(record.route_key || record.id))}`) };
  if (dataset === "resources") return { ...record, kind: "resource", name: record.title, route: String(record.route || record.canonical_path || `/catalog/resources/${encodeURIComponent(String(record.resource_type || "resource"))}/${encodeURIComponent(record.id)}`) };
  return { ...record, kind: "technology", route: String(record.route || `/catalog/technologies/${encodeURIComponent(String(record.slug || record.id))}`) };
}

export async function listCatalog(dataset: CatalogDataset, query: CatalogQuery, signal?: AbortSignal): Promise<CatalogCollection> {
  const params = new URLSearchParams();
  Object.entries(query).forEach(([key, value]) => {
    if (value !== undefined && value !== "") params.set(key, String(value));
  });
  const response = await catalogFetch(`/api/v1/catalog/${dataset}?${params}`, { signal, headers: { Accept: "application/json" }, cache: "default" });
  if (!response.ok) throw new Error(`${dataset} list response ${response.status}`);
  const payload = await response.json() as CatalogCollection;
  if (!Array.isArray(payload.records)) throw new Error(`${dataset} list response did not contain records`);
  return { ...payload, records: payload.records.map((record) => routeRecord(dataset, record)) };
}

async function getCatalogMember(path: string, label: string, signal?: AbortSignal): Promise<CatalogMember> {
  const response = await catalogFetch(path, { signal, headers: { Accept: "application/json" }, cache: "default" });
  if (!response.ok) throw new Error(`${label} member response ${response.status}`);
  const payload = await response.json() as CatalogMember;
  if (!payload.item || typeof payload.item !== "object") throw new Error(`${label} member response 404`);
  return payload;
}

export async function getCatalogRepositoryMember(owner: string, repository: string, signal?: AbortSignal) {
  return getCatalogMember(`/api/v1/catalog/repositories/${encodeURIComponent(owner)}/${encodeURIComponent(repository)}`, "repositories", signal);
}

export async function getCatalogPackageMember(routeKey: string, signal?: AbortSignal) {
  return getCatalogMember(`/api/v1/catalog/packages/${encodeURIComponent(routeKey)}`, "packages", signal);
}

export async function getCatalogResourceMember(resourceType: string, routeKey: string, signal?: AbortSignal) {
  const response = await catalogFetch(`/api/v1/catalog/resources/${encodeURIComponent(resourceType)}/${encodeURIComponent(routeKey)}`, { signal, headers: { Accept: "application/json" }, cache: "default" });
  if (!response.ok) throw new Error(`resources member response ${response.status}`);
  return await response.json() as CatalogMember;
}

export async function getCatalogReleaseMember(routeKey: string, signal?: AbortSignal): Promise<CatalogMember> {
  return getCatalogMember(`/api/v1/catalog/releases/${encodeURIComponent(routeKey)}`, "releases", signal);
}

export async function getCatalogTechnologyMember(slug: string, signal?: AbortSignal) {
  return getCatalogMember(`/api/v1/catalog/technologies/${encodeURIComponent(slug)}`, "technologies", signal);
}
